common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,460 @@
+                    .collect(Collectors.toList());
+
+            Collections.reverse(siblingPairs);
+            siblingPairs.forEach(deque::addFirst);
+
+            // Disallow empty parts
+            parts.addAll(
+                    styledTextParts.stream().filter(part -> !part.isEmpty()).toList());
+        }
+
+        return new StyledText(parts, temporaryWorkaround, clickEvents, hoverEvents);
+    }
+
+    public static StyledText fromString(String codedString) {
+        return new StyledText(
+                StyledTextPart.fromComponentWithPossibleFormattingCodes(codedString, Style.EMPTY, null, Style.EMPTY),
+                List.of(),
+                List.of());
+    }
+
+    // We don't want to expose the actual string to the outside world
+    // If you need to do an operation with this string, implement it as a method
+    public String getString(PartStyle.StyleType type) {
+        if (type == PartStyle.StyleType.FULL) {
+            return ComponentUtils.getCoded(temporaryWorkaround);
+        }
+
+        StringBuilder builder = new StringBuilder();
+
+        PartStyle previousStyle = null;
+        for (StyledTextPart part : parts) {
+            builder.append(part.getString(previousStyle, type));
+            previousStyle = part.getPartStyle();
+        }
+
+        return builder.toString();
+    }
+
+    public MutableComponent getComponent() {
+        if (parts.isEmpty()) {
+            return Component.empty();
+        }
+
+        MutableComponent component = parts.get(0).getComponent();
+
+        for (int i = 1; i < parts.size(); i++) {
+            component.append(parts.get(i).getComponent());
+        }
+
+        return component;
+    }
+
+    public static StyledText join(String separator, StyledText... texts) {
+        List<StyledTextPart> parts = new ArrayList<>();
+        List<ClickEvent> clickEvents = new ArrayList<>();
+        List<HoverEvent> hoverEvents = new ArrayList<>();
+
+        final int length = texts.length;
+        for (int i = 0; i < length; i++) {
+            StyledText text = texts[i];
+            parts.addAll(text.parts);
+
+            if (i != length - 1) {
+                // Possible micro-optimization: If the separator does not have formatting, use the normal constructor
+                parts.addAll(
+                        StyledTextPart.fromComponentWithPossibleFormattingCodes(separator, Style.EMPTY, null, null));
+            }
+
+            clickEvents.addAll(text.clickEvents);
+            hoverEvents.addAll(text.hoverEvents);
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    public static StyledText join(String separator, Iterable<StyledText> texts) {
+        return join(separator, Iterables.toArray(texts, StyledText.class));
+    }
+
+    public static StyledText concat(StyledText... texts) {
+        List<StyledTextPart> parts = new ArrayList<>();
```

Comment:
It could be changed, but I don't really see the point. If you want to make sure it works, unit tests can be used. (Although I think we should add tests after/before your PR, to test everything changed there)